command/agent/template/template.go

```diff
@@ -0,0 +1,186 @@
+// Package template is responsible for rendering user supplied templates to
+// disk. The Server type managing the lifecycle of an internal Consul Template
+// Runner
+package template
+
+import (
+	"context"
+	"strings"
+
+	ctconfig "github.com/hashicorp/consul-template/config"
+	"github.com/hashicorp/consul-template/manager"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/command/agent/config"
+)
+
+// ServerConfig is a config struct for setting up the basic parts of the
+// Server
+type ServerConfig struct {
+	Logger hclog.Logger
+	// Client        *api.Client
+	VaultConf     *config.Vault
+	ExitAfterAuth bool
+
+	Namespace string
+}
+
+// Server manages the Consul Template Runner which renders templates
+type Server struct {
+	// config holds the ServerConfig used to create it. It's passed along in other
+	// methods
+	config *ServerConfig
+
+	// runner is the consul-template runner
+	runner *manager.Runner
+
+	// unblockCh is used to block until a template is rendered
+	unblockCh chan struct{}
+
+	// Templates holds the parsed Consul Templates
+	Templates []*ctconfig.TemplateConfig
+
+	DoneCh        chan struct{}
+	logger        hclog.Logger
+	exitAfterAuth bool
+}
+
+// NewServer returns a new configured server
+func NewServer(conf *ServerConfig) (*Server, chan struct{}) {
```

Comment:
Might be worth thinking about returning a `<- chan struct{}` if we need to enforce that the caller can only read from the channel.

Reply:
I followed the pattern established by AuthHandler and SinkServer:

- https://github.com/hashicorp/vault/blob/a1e3c83368e8ae2fe6d54759c24cff53a74ee901/command/agent.go#L489-L512

In general I agree that if the caller is expected to listen that returning the channel at creation is a good idea, however I think with the limited application here (`template.Server` is only used in Agent, only used here, etc) that it's better to conform to the already established precedent. 

I'll leave this unresolved for a while in case you or others have more feedback here, but I'm inclined to leave it as is.

Reply:
I don't think it can be `<-chan struct{}` since it needs to be `close()`d in the Run method.

Reply:
@tyrannosaurus-becks I totally misread your feedback and responded to something you didn't actually say, sorry! 

You are both correct I think. Jim is correct that you cannot close a receive-only channel, however we don't need to close it on the receiving end so I believe we can use a normal `chan` as a return `<- chan`. I patched that in 429285a92

Reply:
Oh I see. I was looking at `close(ts.unblockCh)` and was considering making the the chan type in the struct itself receive only, which wouldn't work. But yeah, receive only from NewServer is fine.

Reply:
👍 I've made that change, so I'm going to resolve this, thanks!